provider/source/src/time_zones/convert.rs

```diff
@@ -40,500 +32,551 @@ fn type_fallback(zone_format: &ZoneFormat) -> Option<&String> {
-        if let Some(alias) = &bcp47_tzid_data.alias {
-            for data_value in alias.split(' ') {
-                bcp47_tzids.insert(data_value.to_string(), *bcp47_tzid);
-            }
-        }
-    }
-    bcp47_tzids
-}
-
-/// Returns a map from BCP-47 ID to a single canonical long identifier.
-///
-/// For example: "inccu" to "Asia/Kolkata"
-pub(crate) fn compute_canonical_tzids_btreemap(
-    bcp47_tzids_resource: &BTreeMap<TimeZoneBcp47Id, Bcp47TzidAliasData>,
-) -> BTreeMap<TimeZoneBcp47Id, String> {
-    let mut canonical_tzids = BTreeMap::new();
-    for (bcp47_tzid, bcp47_tzid_data) in bcp47_tzids_resource.iter() {
-        if Some(true) == bcp47_tzid_data.deprecated {
-            // skip
-        } else if let Some(iana) = &bcp47_tzid_data.iana {
-            canonical_tzids.insert(*bcp47_tzid, iana.clone());
-        } else if let Some(iana) = &bcp47_tzid_data
-            .alias
-            .as_ref()
-            .and_then(|s| s.split(' ').next())
-        {
-            canonical_tzids.insert(*bcp47_tzid, String::from(*iana));
-        } else {
-            debug_assert!(
-                false,
-                "Could not find canonical IANA for bcp47 time zone: {bcp47_tzid:?}"
-            );
+impl DataProvider<TimeZoneFormatsV1Marker> for SourceDataProvider {
+    fn load(&self, req: DataRequest) -> Result<DataResponse<TimeZoneFormatsV1Marker>, DataError> {
+        self.check_req::<TimeZoneFormatsV1Marker>(req)?;
+
+        let time_zone_names = &self
+            .cldr()?
+            .dates("gregorian")
+            .read_and_parse::<cldr_serde::time_zones::time_zone_names::Resource>(
+                req.id.locale,
+                "timeZoneNames.json",
+            )?
+            .main
+            .value
+            .dates
+            .time_zone_names;
+
+        fn parse_hour_format(hour_format: &str) -> (Cow<'static, str>, Cow<'static, str>) {
+            // e.g. "+HH:mm;-HH:mm" -> ("+HH:mm", "-HH:mm")
+            let index = hour_format.rfind(';').unwrap();
+            let positive = hour_format[0..index].to_owned();
+            let negative = hour_format[index + 1..].to_owned();
+            (Cow::Owned(positive), Cow::Owned(negative))
         }
-    }
-    canonical_tzids
-}
 
-/// Returns a map from metazone long identifier to metazone BCP-47 ID.
-///
-/// For example: "America_Central" to "amce"
-fn compute_meta_zone_ids_btreemap(
-    meta_zone_ids_resource: &BTreeMap<MetazoneId, MetazoneAliasData>,
-) -> BTreeMap<String, MetazoneId> {
-    let mut meta_zone_ids = BTreeMap::new();
-    for (meta_zone_id, meta_zone_id_data) in meta_zone_ids_resource.iter() {
-        meta_zone_ids.insert(meta_zone_id_data.long_id.to_string(), *meta_zone_id);
-    }
-    // TODO(#1781): Remove this special case once the short id is updated in CLDR
-    meta_zone_ids.insert("Yukon".to_owned(), MetazoneId(tinystr!(4, "yuko")));
-    meta_zone_ids
-}
-
-impl From<CldrTimeZonesData<'_>> for TimeZoneFormatsV1<'static> {
```

Comment:
Nit: I prefer the coding style of `impl From<CldrStruct> for Icu4xStruct` because

1. The impl can be used even if not loading data; for example, the processing of some other key might depend on TimeZoneFormatsV1, and it shouldn't need to invoke `impl DataProvider` in order to get the struct. This happens a lot in datetime, for example. I already call `From` impls in some places that need data from elsewhere.
2. More focused code purpose. Keeping `impl DataProvider` small seems like a benefit.
3. Less indentation

Reply:
> The impl can be used even if not loading data; for example, the processing of some other key might depend on TimeZoneFormatsV1, and it shouldn't need to invoke impl DataProvider in order to get the struct. This happens a lot in datetime, for example. I already call From impls in some places that need data from elsewhere.

If that happens we can factor out such code. At the moment there is already shared code that is factored out, but code specific to one data marker isn't.

> Keeping impl DataProvider small seems like a benefit.

Disagree. The data provider implementation should compute the data struct without tons of indirection. The previous code that used a macro and `From` implementations was hard to follow and debug, because cross-references don't work.

Reply:
DataProvider impls on SourceDataProvider should be short and focused: parse the locale, load the JSON it needs, call some conversion function, and return the payload. I find it hard to read SourceDataProvider code whose functions are multi-purpose, long, and deeply nested (10+ levels of indentation!). I prefer if procedural code (data provider impls) and logical code (map from JSON to ICU4X data struct) are separate functions. This should be the default, not only in cases where the code can be shared.

Reply:
Map from JSON to data struct breaks down when the data struct mixes data from x JSON files plus TZDB. Defining intermediate structs like this had before is definitely worse than what I have implemented.